### DISCUSSION TITLE
problem for uploading an old code now

### DISCUSSION THREAD

**Discussion by benwadub** [tttapa/Control-Surface#592] [General]
hi, I wanted to feu-load an old code to my teensy 4.0 but it seems that there is a problem with the encoder library, di you changed something here?
cpp```
In file included from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:98:0,
                 from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:1:
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Inputs/MCU/VU.hpp:188:7: warning: 'CS::MCU::VU' has a field 'CS::MCU::VU::decayTimer' whose type uses the anonymous namespace
 class VU : public MatchingMIDIInputElement<MIDIMessageType::CHANNEL_PRESSURE,
       ^
In file included from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:2:0:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:148:2: error: reference to 'Encoder_internal_state_t' is ambiguous
  Encoder_internal_state_t encoder;
  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:67:3: note: candidates are: typedef struct Encoder_internal_state_t Encoder_internal_state_t
 } Encoder_internal_state_t;
   ^
In file included from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Abstract/MIDIAbsoluteEncoder.hpp:11:0,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/CCAbsoluteEncoder.hpp:3,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:48,
                 from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:1:
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/Encoder.h:48:8: note:                 struct CS::Encoder_internal_state_t
 struct Encoder_internal_state_t {
        ^
In file included from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:2:0:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:153:9: error: reference to 'Encoder_internal_state_t' is ambiguous
  static Encoder_internal_state_t * interruptArgs[ENCODER_ARGLIST_SIZE];
         ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:67:3: note: candidates are: typedef struct Encoder_internal_state_t Encoder_internal_state_t
 } Encoder_internal_state_t;
   ^
In file included from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Abstract/MIDIAbsoluteEncoder.hpp:11:0,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/CCAbsoluteEncoder.hpp:3,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:48,
                 from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:1:
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/Encoder.h:48:8: note:                 struct CS::Encoder_internal_state_t
 struct Encoder_internal_state_t {
        ^
In file included from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:2:0:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:206:21: error: reference to 'Encoder_internal_state_t' is ambiguous
  static void update(Encoder_internal_state_t *arg) {
                     ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:67:3: note: candidates are: typedef struct Encoder_internal_state_t Encoder_internal_state_t
 } Encoder_internal_state_t;
   ^
In file included from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Abstract/MIDIAbsoluteEncoder.hpp:11:0,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/CCAbsoluteEncoder.hpp:3,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:48,
                 from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:1:
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/Encoder.h:48:8: note:                 struct CS::Encoder_internal_state_t
 struct Encoder_internal_state_t {
        ^
In file included from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:2:0:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:206:21: error: 'Encoder_internal_state_t' has not been declared
  static void update(Encoder_internal_state_t *arg) {
                     ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:379:47: error: reference to 'Encoder_internal_state_t' is ambiguous
  static uint8_t attach_interrupt(uint8_t pin, Encoder_internal_state_t *state) {
                                               ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:67:3: note: candidates are: typedef struct Encoder_internal_state_t Encoder_internal_state_t
 } Encoder_internal_state_t;
   ^
In file included from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Abstract/MIDIAbsoluteEncoder.hpp:11:0,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/CCAbsoluteEncoder.hpp:3,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:48,
                 from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:1:
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/Encoder.h:48:8: note:                 struct CS::Encoder_internal_state_t
 struct Encoder_internal_state_t {
        ^
In file included from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:2:0:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:379:47: error: 'Encoder_internal_state_t' has not been declared
  static uint8_t attach_interrupt(uint8_t pin, Encoder_internal_state_t *state) {
                                               ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In constructor 'Encoder::Encoder(uint8_t, uint8_t)':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:82:3: error: 'encoder' was not declared in this scope
   encoder.pin1_register = PIN_TO_BASEREG(pin1);
   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In member function 'int32_t Encoder::read()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:107:12: error: 'encoder' was not declared in this scope
    update(&encoder);
            ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:111:17: error: 'encoder' was not declared in this scope
   int32_t ret = encoder.position;
                 ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In member function 'int32_t Encoder::readAndReset()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:118:12: error: 'encoder' was not declared in this scope
    update(&encoder);
            ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:122:17: error: 'encoder' was not declared in this scope
   int32_t ret = encoder.position;
                 ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In member function 'void Encoder::write(int32_t)':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:129:3: error: 'encoder' was not declared in this scope
   encoder.position = p;
   ^
In file included from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/Encoder.h:36:0,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Abstract/MIDIAbsoluteEncoder.hpp:11,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/CCAbsoluteEncoder.hpp:3,
                 from /Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:48,
                 from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:1:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::update(int*)':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:292:40: error: request for member 'pin1_register' in '* arg', which is of non-class type 'int'
   uint8_t p1val = DIRECT_PIN_READ(arg->pin1_register, arg->pin1_bitmask);
                                        ^
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/utility/direct_pin_read.h:23:46: note: in definition of macro 'DIRECT_PIN_READ'
 #define DIRECT_PIN_READ(base, mask)     (((*(base)) & (mask)) ? 1 : 0)
                                              ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:292:60: error: request for member 'pin1_bitmask' in '* arg', which is of non-class type 'int'
   uint8_t p1val = DIRECT_PIN_READ(arg->pin1_register, arg->pin1_bitmask);
                                                            ^
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/utility/direct_pin_read.h:23:56: note: in definition of macro 'DIRECT_PIN_READ'
 #define DIRECT_PIN_READ(base, mask)     (((*(base)) & (mask)) ? 1 : 0)
                                                        ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:293:40: error: request for member 'pin2_register' in '* arg', which is of non-class type 'int'
   uint8_t p2val = DIRECT_PIN_READ(arg->pin2_register, arg->pin2_bitmask);
                                        ^
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/utility/direct_pin_read.h:23:46: note: in definition of macro 'DIRECT_PIN_READ'
 #define DIRECT_PIN_READ(base, mask)     (((*(base)) & (mask)) ? 1 : 0)
                                              ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:293:60: error: request for member 'pin2_bitmask' in '* arg', which is of non-class type 'int'
   uint8_t p2val = DIRECT_PIN_READ(arg->pin2_register, arg->pin2_bitmask);
                                                            ^
/Users/benoitparisot/Documents/Arduino/libraries/Control-Surface-master/src/Submodules/Encoder/utility/direct_pin_read.h:23:56: note: in definition of macro 'DIRECT_PIN_READ'
 #define DIRECT_PIN_READ(base, mask)     (((*(base)) & (mask)) ? 1 : 0)
                                                        ^
In file included from /Users/benoitparisot/Library/projets arduino/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons/controleur_BEN_and_ED_ableton9_16_avec_display_et_boutons.ino:2:0:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:294:24: error: request for member 'state' in '* arg', which is of non-class type 'int'
   uint8_t state = arg->state & 3;
                        ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:297:8: error: request for member 'state' in '* arg', which is of non-class type 'int'
   arg->state = (state >> 2);
        ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:300:10: error: request for member 'position' in '* arg', which is of non-class type 'int'
     arg->position++;
          ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:303:10: error: request for member 'position' in '* arg', which is of non-class type 'int'
     arg->position--;
          ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:306:10: error: request for member 'position' in '* arg', which is of non-class type 'int'
     arg->position += 2;
          ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:309:10: error: request for member 'position' in '* arg', which is of non-class type 'int'
     arg->position -= 2;
          ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static uint8_t Encoder::attach_interrupt(uint8_t, int*)':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:383:5: error: 'interruptArgs' was not declared in this scope
     interruptArgs[0] = state;
     ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr0()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:752:34: error: 'interruptArgs' was not declared in this scope
  static void isr0(void) { update(interruptArgs[0]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr1()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:755:34: error: 'interruptArgs' was not declared in this scope
  static void isr1(void) { update(interruptArgs[1]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr2()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:758:34: error: 'interruptArgs' was not declared in this scope
  static void isr2(void) { update(interruptArgs[2]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr3()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:761:34: error: 'interruptArgs' was not declared in this scope
  static void isr3(void) { update(interruptArgs[3]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr4()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:764:34: error: 'interruptArgs' was not declared in this scope
  static void isr4(void) { update(interruptArgs[4]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr5()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:767:34: error: 'interruptArgs' was not declared in this scope
  static void isr5(void) { update(interruptArgs[5]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr6()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:770:34: error: 'interruptArgs' was not declared in this scope
  static void isr6(void) { update(interruptArgs[6]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr7()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:773:34: error: 'interruptArgs' was not declared in this scope
  static void isr7(void) { update(interruptArgs[7]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr8()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:776:34: error: 'interruptArgs' was not declared in this scope
  static void isr8(void) { update(interruptArgs[8]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr9()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:779:34: error: 'interruptArgs' was not declared in this scope
  static void isr9(void) { update(interruptArgs[9]); }
                                  ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr10()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:782:35: error: 'interruptArgs' was not declared in this scope
  static void isr10(void) { update(interruptArgs[10]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr11()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:785:35: error: 'interruptArgs' was not declared in this scope
  static void isr11(void) { update(interruptArgs[11]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr12()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:788:35: error: 'interruptArgs' was not declared in this scope
  static void isr12(void) { update(interruptArgs[12]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr13()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:791:35: error: 'interruptArgs' was not declared in this scope
  static void isr13(void) { update(interruptArgs[13]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr14()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:794:35: error: 'interruptArgs' was not declared in this scope
  static void isr14(void) { update(interruptArgs[14]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr15()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:797:35: error: 'interruptArgs' was not declared in this scope
  static void isr15(void) { update(interruptArgs[15]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr16()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:800:35: error: 'interruptArgs' was not declared in this scope
  static void isr16(void) { update(interruptArgs[16]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr17()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:803:35: error: 'interruptArgs' was not declared in this scope
  static void isr17(void) { update(interruptArgs[17]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr18()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:806:35: error: 'interruptArgs' was not declared in this scope
  static void isr18(void) { update(interruptArgs[18]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr19()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:809:35: error: 'interruptArgs' was not declared in this scope
  static void isr19(void) { update(interruptArgs[19]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr20()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:812:35: error: 'interruptArgs' was not declared in this scope
  static void isr20(void) { update(interruptArgs[20]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr21()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:815:35: error: 'interruptArgs' was not declared in this scope
  static void isr21(void) { update(interruptArgs[21]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr22()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:818:35: error: 'interruptArgs' was not declared in this scope
  static void isr22(void) { update(interruptArgs[22]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr23()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:821:35: error: 'interruptArgs' was not declared in this scope
  static void isr23(void) { update(interruptArgs[23]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr24()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:824:35: error: 'interruptArgs' was not declared in this scope
  static void isr24(void) { update(interruptArgs[24]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr25()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:827:35: error: 'interruptArgs' was not declared in this scope
  static void isr25(void) { update(interruptArgs[25]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr26()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:830:35: error: 'interruptArgs' was not declared in this scope
  static void isr26(void) { update(interruptArgs[26]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr27()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:833:35: error: 'interruptArgs' was not declared in this scope
  static void isr27(void) { update(interruptArgs[27]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr28()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:836:35: error: 'interruptArgs' was not declared in this scope
  static void isr28(void) { update(interruptArgs[28]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr29()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:839:35: error: 'interruptArgs' was not declared in this scope
  static void isr29(void) { update(interruptArgs[29]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr30()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:842:35: error: 'interruptArgs' was not declared in this scope
  static void isr30(void) { update(interruptArgs[30]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr31()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:845:35: error: 'interruptArgs' was not declared in this scope
  static void isr31(void) { update(interruptArgs[31]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr32()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:848:35: error: 'interruptArgs' was not declared in this scope
  static void isr32(void) { update(interruptArgs[32]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr33()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:851:35: error: 'interruptArgs' was not declared in this scope
  static void isr33(void) { update(interruptArgs[33]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr34()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:854:35: error: 'interruptArgs' was not declared in this scope
  static void isr34(void) { update(interruptArgs[34]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr35()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:857:35: error: 'interruptArgs' was not declared in this scope
  static void isr35(void) { update(interruptArgs[35]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr36()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:860:35: error: 'interruptArgs' was not declared in this scope
  static void isr36(void) { update(interruptArgs[36]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr37()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:863:35: error: 'interruptArgs' was not declared in this scope
  static void isr37(void) { update(interruptArgs[37]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr38()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:866:35: error: 'interruptArgs' was not declared in this scope
  static void isr38(void) { update(interruptArgs[38]); }
                                   ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h: In static member function 'static void Encoder::isr39()':
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder/Encoder.h:869:35: error: 'interruptArgs' was not declared in this scope
  static void isr39(void) { update(interruptArgs[39]); }
                                   ^```
I got these error

**Comment by Veroledez** [comment]
Hola, publica el codigo completo. Pero parece que se debe a la biblioteca encoder, ahora control surface no nececita de esa libreria y tal vez te este creando conflicto, ademas han cambiado algunos constructores de objetos, por ahi tambien podria haber algo.
Mi consejo borra la libreria encoder y ve si da fallo, si aun continua tienes que revisar tu codigo y compararlo con los ejemplos nuevos haber que cambio

**Comment by tttapa** [comment]
Please see the [change log](https://github.com/tttapa/Control-Surface#change-log-and-updating), you should not include Encoder.h when using the 2.x version, it will clash with the included encoder code.

**Reply by benwadub** [reply]
Thanks I ll try to update my code I haven t s’en the changes!

**Reply by benwadub** [reply]
ok so I can now upload my code but my controler seems to not send midi data! it is recognized by my computer with his name but nothing is outputting from the usb midi interface! the hardware was working nicely since more than a year

**Reply by benwadub** [reply]
it seems that it s my display that is buggy, if I remove the part of the code that display my logo at startup all is working!
